engines/sci/graphics/palette32.cpp

```diff
@@ -21,20 +21,156 @@
  */
 
 #include "common/file.h"
+#include "common/system.h"
+#include "graphics/palette.h"
 
 #include "sci/sci.h"
+#include "sci/resource.h"
 #include "sci/graphics/palette32.h"
+#include "sci/graphics/screen.h"
 
 namespace Sci {
 	
 GfxPalette32::GfxPalette32(ResourceManager *resMan, GfxScreen *screen)
-	: GfxPalette(resMan, screen), _clutTable(0), _cyclers(), _cycleMap() {}
+	: GfxPalette(resMan, screen),
+	_clutTable(nullptr),
+	// Palette cycling
+	_cyclers(), _cycleMap(),
+	// Palette varying
+	_varyTime(0), _varyDirection(0), _varyTargetPercent(0),
+	_varyTargetPalette(nullptr), _varyStartPalette(nullptr),
+	_varyFromColor(0), _varyToColor(255), _varyNumTimesPaused(0),
+	_varyPercent(_varyTargetPercent), _varyLastTick(0),
+	// Palette versioning
+	_version(1), _needsUpdate(false) {
+		// TODO: Sci::Palette should have a default constructor that does
+		// all of this
+		for (int color = 0; color < ARRAYSIZE(_sysPalette.colors); color++) {
+			_sourcePalette.colors[color] = _nextPalette.colors[color] = _sysPalette.colors[color];
+			// TODO: Intensity and mapping are not used by SCI32 but are used by
+			// other parts of the graphic engine right now. Replace those parts.
+			_sourcePalette.intensity[color] = _nextPalette.intensity[color] = 100;
+			_sourcePalette.mapping[color] = _nextPalette.mapping[color] = color;
+		}
+
+		for (int i = 0; i < ARRAYSIZE(_fadeTable); ++i) {
+			_fadeTable[i] = 100;
+		}
+
+		// TODO: In SCI engine, GfxPalette (PaletteMgr) constructor handles
+		// loading the default palette, but ScummVM already loads it from
+		// somewhere else, so should ScummVM be updated to match SCI or should
+		// it be left the way it is already?
+		assert(kernelSetFromResource(999, true));
+	}
 
 GfxPalette32::~GfxPalette32() {
 	unloadClut();
+	varyOff();
 	cycleAllOff();
 }
 
+void GfxPalette32::submit(Palette &palette) {
+	// TODO: This code won’t do much unless/until submit can
+	// receive HunkPalettes in addition to SOLPalettes. In SCI engine,
+	// SOLPalettes do not have timestamp/version fields, only
+	// HunkPalettes do, but in ScummVM there is only one type of palette.
+	if (palette.timestamp == _version) {
```

Comment:
Is there a reason why we would want to introduce another type of palette? If not, we can just remove this TODO. Otherwise, we might want to differentiate the two (but personally, I don't see the reason why we would want to do that...)